build.sbt

```diff
@@ -61,9 +62,38 @@ def removeScalacOptionsInTest(scalaVersion: String) =
     case _ => Seq("-Wvalue-discard", "-Wunused:explicits", "-Wunused:params", "-Wunused:imports")
   }
 
+def macrosDepsByVersion(scalaVersion: String) =
+  CrossVersion.partialVersion(scalaVersion) match {
+    case Some((2, _)) =>
+      Seq(
+        "com.chuusai" %% "shapeless" % shapelessVersion,
+      )
+    case Some((3, _)) => 
+      Seq("org.typelevel" %% "shapeless3-deriving" % shapeless3Version)
+    case _ => Seq.empty  
+
+  }
+
+def maybeReflect(scalaVersion: String) =
+  CrossVersion.partialVersion(scalaVersion) match {
+    case Some((2, _)) =>
+      Seq(
+        "org.scala-lang" % "scala-reflect" % scalaVersion
+      )
+    case _ => Seq.empty  
+  }
+
+def compileOptsByVersion(scalaVersion: String) =
+  CrossVersion.partialVersion(scalaVersion) match {
+    case Some((2, _)) =>
+      Seq("-Ywarn-macros:after")
+    case _ => 
+      Seq.empty
+  }  
```

Comment:
What about adding to this the `enablePartialUnificationIn2_12` logic I have below?

Reply:
Sure, sounds good to me.

Reply:
I won't address this one as `enablePartialUnificationIn2_12` is only currently used in the `catsEffect` project to where as `compileOptsByVersion` is used as part of `commonSettings`.